src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs

```diff
@@ -66,16 +69,45 @@ public ServiceEndpoint(string connectionString, EndpointType type = EndpointType
             Name = name;
         }
 
-        public ServiceEndpoint(ServiceEndpoint endpoint)
+        /// <summary>
+        /// Aad constructor with key
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="endpoint"></param>
+        /// <param name="credential"></param>
+        public ServiceEndpoint(string key, Uri endpoint, TokenCredential credential) : this(endpoint, credential)
+        {
+            (Name, EndpointType) = ParseKey(key);
+        }
+
+        /// <summary>
+        /// Aad constructor
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <param name="credential"></param>
+        public ServiceEndpoint(Uri endpoint, TokenCredential credential = null)
```

Comment:
I think we should ensure credential is not null. The fallback login of `TokenCredential` should be handled by the callee.